### PR TITLE
Add CSP Injection to username XSS challenge

### PR DESCRIPTION
**Pull request by Scar26**
Fixes(?): #1245 

The challenge 'Classic Stored XSS' now requires the user to perform a CSP injection through the profile image url and bypass the regex check on username in order to perform XSS.

Bumped up the challenge difficulty from two to four

P.S : Users can now submit any URL for their profile images. This new feature is pretty vulnerable as of now and can be used in future challenges

### CODE REVIEW

**Comment by Scar26** [comment]
The travis build seems to be failing before it begins...
tests seem to be running fine locally

**Comment by J12934** [comment]
@Scar26 I‘ve restarted the build.

**Comment by Scar26** [comment]
@J12934 It's glitching again, the build has been queued for 20 minutes

**Comment by bkimminich** [comment]
Develop branch is building your profile logo fix, and I think branch builds take precedence over PR builds on Travis. But your PR build started 2min ago with Linting... 🐌

**Comment by Scar26** [comment]
It's throwing an error on the feedbackApiSpec test which I haven't even changed
It's passing on local : 
feedbackapitest

**Comment by Scar26** [comment]
Just pushed the same commit again and it passed by itself :p

**Comment by Scar26** [comment]
@bkimminich @J12934 can you please review this PR and let me know if there are any changes to be made

**Comment by J12934** [comment]
Will try to take a deeper look tomorrow

**Comment by Scar26** [comment]
@J12934, just a reminder :)

File: routes/userProfile.js [hunk 1]
```diff
@@ -43,6 +44,20 @@ module.exports = function getUserProfile () {
           template = template.replace(/_primDark_/g, theme.primDark)
           template = template.replace(/_logo_/g, utils.extractFilename(config.get('application.logo')))
           const fn = pug.compile(template)
+          let CSP
+          if (user.dataValues.profileImage.match(/^https:/) !== null) {
+            CSP = `img-src 'self' ${user.dataValues.profileImage}; script-src 'self' 'unsafe-eval' https://code.getmdl.io http://ajax.googleapis.com`
+            if (!utils.disableOnContainerEnv()) {
```
Comment (J12934) [inline routes/userProfile.js]: Does it really make sense to disable this on containerized environments?
We normally just disable remote code execution type challenge on docker.

File: models/user.js [hunk 2]
```diff
@@ -55,7 +54,7 @@ module.exports = (sequelize, { STRING, BOOLEAN }) => {
     },
     profileImage: {
       type: STRING,
-      defaultValue: 'default.svg'
+      defaultValue: '/assets/public/images/uploads/default.svg'
```
Comment (J12934) [inline models/user.js:57]: Was this broken before?
Or where is this change coming from?

File: routes/profileImageRawUrlUpload.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2014-2020 Bjoern Kimminich.
+ * SPDX-License-Identifier: MIT
+ */
+
+const models = require('../models/index')
+const insecurity = require('../lib/insecurity')
+
+module.exports = function profileImageURLUpdate () {
```
Comment (J12934) [inline routes/profileImageRawUrlUpload.js]: Not a big fan of basically having two sepearate profilePic url routes.
Is there a way to to combine them both in one?

File: test/e2e/profileSpec.js [hunk 4]
```diff
@@ -7,7 +7,7 @@ const config = require('config')
 const utils = require('../../lib/utils')
 
 describe('/profile', () => {
-  let username, submitButton, url, setButton
+  let username, submitButton, url, setButton, profileImage, sendButton
```
Comment (J12934) [inline test/e2e/profileSpec.js]: I think these vars should be renamed to reflect the texts of the buttons 😅
If i had to pick which of the `submitButton`, `setButton` or `sendButtons` did what I'd have not the slightest clue.

**Review by J12934 — COMMENTED** [review]
Nice challenge idea 👍

What I don't really like about this is having two separate input fields and two separate routes for the profilePic urls. Would be cool to find a solution which would work for the CSP bypass XSS Challenge and the SSRF challenge. Or maybe move the SSRF challenge to another part of the application?

File: routes/userProfile.js (hunk 1 above)
Comment (Scar26) [reply routes/userProfile.js]: Sorry copied this part from the other route. Will change this ASAP

File: models/user.js (hunk 2 above)
Comment (Scar26) [reply models/user.js:57]: originally if you added a profile pic via URL, it was also saved as a file in the assets folder and then referenced via it's designated route on loading time. I changed the user model to include the complete path to the image. Changing it to this way seemed like a neat way to make sure they are referenced directly from the url and I don't have to add extra parameters to classify files,gravatar links and general urls

File: routes/profileImageRawUrlUpload.js (hunk 3 above)
Comment (Scar26) [reply routes/profileImageRawUrlUpload.js]: Will do!

File: test/e2e/profileSpec.js (hunk 4 above)
Comment (Scar26) [reply test/e2e/profileSpec.js]: Sorry about that, will change this ASAP

File: routes/userProfile.js (hunk 1 above)
Comment (Scar26) [reply routes/userProfile.js]: Sorry this wasn't copied... not sure why I kept this

Comment (Scar26) [reply routes/userProfile.js]: Done

File: routes/profileImageRawUrlUpload.js (hunk 3 above)
Comment (Scar26) [reply routes/profileImageRawUrlUpload.js]: Done

File: test/e2e/profileSpec.js (hunk 4 above)
Comment (Scar26) [reply test/e2e/profileSpec.js]: Done

**Comment by Scar26** [comment]
Both challenges are now accessible via the same input field and route.

Though the username xss challenge now requires a bit more fuzzing on the player's part to notice that  they have to enter an invalid URL for it to be directly added to the CSP

Sorry for the delay, I had to be away for a few days

**Comment by Scar26** [comment]
Closing because of too much clutter.

I will thoroughly ensure it's all fixed and will open a new PR shortly. Apologies for the delay and confusion in this one, wasn't able to get much work done these past few days

**Pull request closed without merging**